pkg/reconciler/pipeline/dag/dag.go

```diff
@@ -66,16 +66,29 @@ func (g *Graph) addPipelineTask(t Task) (*Node, error) {
 	return newNode, nil
 }
 
-// Build returns a valid pipeline Graph. Returns error if the pipeline is invalid
-func Build(tasks Tasks) (*Graph, error) {
+// BuildWithoutLinks builds a Pipeline Graph with the specified tasks as nodes (without any links).
+// Returns error if the pipeline is invalid
```

Comment:
I don't understand what BuildWithoutLinks is doing, could you add more detail, e.g. explain what "links" means in this context? I think it means that the graph is created without any edges, in which case it's not a graph (i mean technically it is but not in a useful way) - is the idea to get a list of nodes? would it make sense to create a different data structure or maybe just return a list of []Node?

Reply:
`BuildWithoutLinks` is creating graph without linking any dependencies. 

Task results introduces dependency between two tasks i.e. consumer task is dependent on producer task and from the `dag` perspective, both tasks are part of the same graph. But in case of `finally` consuming results from `dag`, both are part of a separate sections. And as per the `finally` contract, it can not depend on any `dag` task.

You are right practically its just a list of `Node`s, all finally tasks being root nodes of a graph.

We could convert it into a list instead of graph. This was originally designed to be a graph to support finally tasks being dependent on (order and resource dependency) other finally tasks.

I have renamed `BuildWithoutLinks` to `BuildWithoutDeps` if its easier to understand.

Reply:
Another alternative is to have a build function taking both `dag` and `finally` and return two graphs:

```
func Build(dag Tasks, finally Tasks) (*Graph, *Graph, error) {
```

Build graph as is for `dag` tasks but while building graph for `finally`, avoid any dependencies if its a `dag` task.

Or

Build graph with all the `tasks` but avoid adding dependencies if it exists in a list of `noDeps`:

```
func Build(tasks Tasks, noDeps Tasks) (*Graph, error) {
```

Call it like:

```
d := Build(dagTasks, []PipelineTasks)
dfinally := Build(finallyTasks, dagTasks)
```

Reply:
Thanks for explaining!

I think it's important to use the right data structure for whatever we're doing - it makes the code much clearer and easier to maintain in the long run.

In this case, if a list of nodes is what we need, do we need to involve a graph object at all? Can we create a function that is not related to Graph at all and produces a list of Nodes from a list of Tasks?

e.g. 

```
func GetNodes(tasks Tasks) []Node
```

Alternatively: maybe the problem is that determining the deps shouldnt be part of building the graph? what if `deps` was passed into the build function?

Reply:
> I think it's important to use the right data structure for whatever we're doing - it makes the code much clearer and easier to maintain in the long run.

yup, building `Graph` for `finally` section is essential to support `runAfter` and `workspaces` in future.

Reply:
> Alternatively: maybe the problem is that determining the deps shouldnt be part of building the graph? what if deps was passed into the build function?

yup, working on this alternative.

Reply:
Taking `deps` out of `build` does simplify `build`, see changes [here](https://github.com/tektoncd/pipeline/compare/master...pritidesai:build-deps?expand=1) 

And, irrespective of `finally` section being built as a `Graph` or a list of `Node[]`, it has to fetch `deps` to validate validate result reference belongs to a valid `dag` task. 

PTAL 🙏

Reply:
Nice! I wonder if it would help if there was a function you could call to get all the deps as a map?

e.g. something like

```

type PipelineTasks []PipelineTask
func (tasks PipelineTasks) Deps() map[string][] {
        deps := map[string][]string{}
	for _, pt := range tasks {
		deps[pt.HashKey()] = pt.Deps()
	}
        return deps
}
```

then the original line:

```
if _, err := dag.Build(PipelineTaskList(tasks)); err != nil {
```

would just become something like :

```
if _, err := dag.Build(PipelineTaskList(tasks),  tasks.Deps()); err != nil {
```

Reply:
(oh! `PipelineTaskList` already IS the type I defined above as `PipelineTasks` :D)

Reply:
Great suggestion, it looks much simpler now, PR # https://github.com/tektoncd/pipeline/pull/3583

Reply:
> validate the result reference belonging to a valid dag task which is missing from the changes in this PR 😢

I have added validation [here](https://github.com/tektoncd/pipeline/pull/3242/files#diff-78ff878240189b20e0ef82f2912c93c0b3c7b445865e04fa9b98aab851be6130R343).